### PR TITLE
Starcoder2 model

### PR DESCRIPTION
**Pull request by jlamypoirier**
The Starcoder2 model, adapted from Mistral. All changes are done through options, so Mistral itself is still supported. Main changes:
* Use layer norm (RMS still available as option)
* Use standard MLP (gated still available as option)
* Add back biases (optional)
* Change (default?) tokenizer class
*Embedding and residual dropout

It does not support absolute embeddings, so can't support Santacoder or Starcoder

Todo:
* Forward changes from #27931, #29027 (and future changes from Feb. 19)
* Documentation
* Copyright
* Point to starcoder2 checkpoint
* Other minor things (see todos)

@younesbelkada

### CODE REVIEW

File: src/transformers/models/starcoder2/modeling_starcoder2.py [hunk 1]
```diff
@@ -0,0 +1,1432 @@
...
+            )
+        else:
+            # 4d mask is passed through the layers
+            attention_mask = _prepare_4d_causal_attention_mask(
+                attention_mask,
+                (batch_size, seq_length),
+                inputs_embeds,
+                past_key_values_length,
+                sliding_window=self.config.sliding_window,
+            )
```
Comment (ArthurZucker) [inline src/transformers/models/starcoder2/modeling_starcoder2.py:1062]: see the new Llama code for this which was simpliefied. I'd rather we take it directly for the attention 😉

**Review by ArthurZucker — COMMENTED** [review]
🔥 looks very good! 
- Let's not support mistral 
- Let's try to take in the new API from #28937 and #27931 
The rest is pretty much alright!

File: src/transformers/models/starcoder2/modeling_starcoder2.py [hunk 2]
```diff
@@ -0,0 +1,1432 @@
...
+
+    def forward(self, hidden_states: Optional[Tuple[torch.FloatTensor]]) -> torch.FloatTensor:
+        hidden_states = self.c_fc(hidden_states)
+        hidden_states = self.act(hidden_states)
+        hidden_states = self.c_proj(hidden_states)
+        hidden_states = nn.functional.dropout(hidden_states, p=self.residual_dropout, training=self.training)
+        return hidden_states
+
+
+class Starcoder2GatedMLP(nn.Module):
```
[the remark this answers is not in the document]
Comment (jlamypoirier) [reply src/transformers/models/starcoder2/modeling_starcoder2.py:182]: It has small changes (bias + dropout I think)

[the remark this answers is not in the document]
Comment (RaymondLi0) [reply src/transformers/models/starcoder2/modeling_starcoder2.py:182]: Should we remove the copied mention from all the classes/methods where we added dropout?
https://github.com/huggingface/transformers/blob/7fac7d80e611c0b01236614942f440c179a7bb5a/src/transformers/models/starcoder2/modeling_starcoder2.py#L262
https://github.com/huggingface/transformers/blob/7fac7d80e611c0b01236614942f440c179a7bb5a/src/transformers/models/starcoder2/modeling_starcoder2.py#L346
https://github.com/huggingface/transformers/blob/7fac7d80e611c0b01236614942f440c179a7bb5a/src/transformers/models/starcoder2/modeling_starcoder2.py#L974

[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/starcoder2/modeling_starcoder2.py:182]: yes otherwise the `check-copies` will fail 😉

**Comment by RaymondLi0** [comment]
I re-created a PR here since Joel is on vacation: https://github.com/huggingface/transformers/pull/29215

**Comment by ArthurZucker** [comment]
Closing as #29215 was merged and starcoder 2 is officially supported
